beacon-chain/rpc/eth/beacon/handlers_validator.go

```diff
@@ -0,0 +1,375 @@
+
+	resp := &GetValidatorResponse{
+		Data:                container,
+		ExecutionOptimistic: isOptimistic,
+		Finalized:           isFinalized,
+	}
+	http2.WriteJson(w, resp)
+}
+
+// GetValidatorBalances returns a filterable list of validator balances.
+func (bs *Server) GetValidatorBalances(w http.ResponseWriter, r *http.Request) {
+	ctx, span := trace.StartSpan(r.Context(), "beacon.GetValidatorBalances")
+	defer span.End()
+
+	stateId := mux.Vars(r)["state_id"]
+	if stateId == "" {
+		http2.HandleError(w, "state_id is required in URL params", http.StatusBadRequest)
+		return
+	}
+	st, err := bs.Stater.State(ctx, []byte(stateId))
+	if err != nil {
+		shared.WriteStateFetchError(w, err)
+		return
+	}
+
+	isOptimistic, err := helpers.IsOptimistic(ctx, []byte(stateId), bs.OptimisticModeFetcher, bs.Stater, bs.ChainInfoFetcher, bs.BeaconDB)
+	if err != nil {
+		http2.HandleError(w, "Could not check optimistic status: "+err.Error(), http.StatusInternalServerError)
+		return
+	}
+	blockRoot, err := st.LatestBlockHeader().HashTreeRoot()
+	if err != nil {
+		http2.HandleError(w, "Could not calculate root of latest block header: "+err.Error(), http.StatusInternalServerError)
+		return
+	}
+	isFinalized := bs.FinalizationFetcher.IsFinalized(ctx, blockRoot)
+
+	rawIds := r.URL.Query()["id"]
+	ids, ok := decodeIds(w, st, rawIds, true /* ignore unknown */)
+	if !ok {
+		return
+	}
+	// return no data if all IDs are ignored
+	if len(rawIds) > 0 && len(ids) == 0 {
+		resp := &GetValidatorBalancesResponse{
+			Data:                []*ValidatorBalance{},
+			ExecutionOptimistic: isOptimistic,
+			Finalized:           isFinalized,
+		}
+		http2.WriteJson(w, resp)
+		return
+	}
+
+	bals := st.Balances()
+	var valBalances []*ValidatorBalance
+	if len(ids) == 0 {
+		valBalances = make([]*ValidatorBalance, len(bals))
+		for i, b := range bals {
+			valBalances[i] = &ValidatorBalance{
+				Index:   strconv.FormatUint(uint64(i), 10),
+				Balance: strconv.FormatUint(b, 10),
+			}
+		}
+	} else {
+		valBalances = make([]*ValidatorBalance, len(ids))
+		for i, id := range ids {
+			valBalances[i] = &ValidatorBalance{
+				Index:   strconv.FormatUint(uint64(id), 10),
+				Balance: strconv.FormatUint(bals[id], 10),
+			}
+		}
+	}
+
+	resp := &GetValidatorBalancesResponse{
+		Data:                valBalances,
+		ExecutionOptimistic: isOptimistic,
+		Finalized:           isFinalized,
+	}
+	http2.WriteJson(w, resp)
+}
+
+// decodeIds takes in a list of validator ID strings (as either a pubkey or a validator index)
+// and returns the corresponding validator indices. It can be configured to ignore well-formed but unknown indices.
+func decodeIds(w http.ResponseWriter, st state.BeaconState, rawIds []string, ignoreUnknown bool) ([]primitives.ValidatorIndex, bool) {
+	ids := make([]primitives.ValidatorIndex, 0, len(rawIds))
+	numVals := uint64(st.NumValidators())
+	for _, rawId := range rawIds {
+		pubkey, err := hexutil.Decode(rawId)
+		if err == nil {
```

Comment:
is it supposed to do something if it errors? didn't see error handling on this.

Reply:
We try to decode pubkey first, if it fails we try to decode a validator index
```
index, err := strconv.ParseUint(rawId, 10, 64)
		if err != nil {
			http2.HandleError(w, fmt.Sprintf("Invalid validator index %s", rawId), http.StatusBadRequest)
			return nil, false
		}
```